idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,1160 @@
+        dfs_sum_revenue(set_idx = 0, indices = [], bidding_indices = [])
+
+    def compute_bids(self, price_forecasts, date, hour):
+
+        '''
+        Solve the model to bid into the markets. After solving, record the bids
+        from the solve.
+
+        Arguments:
+            price_forecasts: price forecasts needed to solve the bidding problem.
+                            {gen: {LMP scenario: [forecast timeseries] }}
+            date: current simulation date
+            hour: current simulation hour
+
+        Returns:
+            None
+        '''
+
+        # update the price forecasts
+        self._pass_price_forecasts(price_forecasts)
+        self.solver.solve(self.model,tee=True)
+        bids = self._assemble_bids()
+        self.record_bids(bids, date = date, hour = hour)
+
+        ## TODO: when to update bidding model??
+
+        return bids
+
+    def _pass_price_forecasts(self, price_forecasts):
+
+        '''
+        Pass the price forecasts into model parameters.
+
+        Arguments:
+            price_forecasts: price forecasts needed to solve the bidding problem.
+                             {generator: {LMP scenario: [forecast timeseries] }}
+
+        Returns:
+            None
+        '''
+
+        model_sets = self.bidding_model_object.indices
+        energy_price = self.model.energy_price
+
+        aux_dict = {'Generators': None, 'Time': None, 'LMP Scenarios': None}
+
+        for idx1 in self.bidding_set[0]:
+            for idx2 in self.bidding_set[1]:
+                for idx3 in self.bidding_set[2]:
+
+                    aux_dict[model_sets[self.bidding_set[0]]] = idx1
+                    aux_dict[model_sets[self.bidding_set[1]]] = idx2
+                    aux_dict[model_sets[self.bidding_set[2]]] = idx3
+
+                    energy_price[idx1,idx2,idx3] = price_forecasts[aux_dict['Generators']][aux_dict['LMP Scenarios']][aux_dict['Time']]
+
+        return
+
+    def _assemble_bids(self):
+
+        '''
+        This methods extract the bids out of the stochastic programming model and
+        organize them.
+
+        Arguments:
+
+        Returns:
+            bids: the bid we computed. It is a dictionary that has this structure
+            {t: {gen:{power: cost}}}.
+        '''
+
+        model_sets = self.bidding_model_object.indices
+        model_sets_inverse = {}
+        for s, n in model_sets.items():
+            model_sets_inverse[n] = s
+
+        # unpack things
+        model_sets = self.bidding_model_object.indices
+        set_ls = list(model_sets.keys())
+
+        power_output = self.bidding_model_object.power_output
+        energy_price = self.model.energy_price
+
+        bids = {}
+        def dfs_extract_bids(set_idx, indices, bidding_indices, t, gen):
+
+            # traveled all the sets, and now we can add the objective
+            if set_idx == len(set_ls):
+                power = round(pyo.value(power_output[tuple(indices)]),2)
```

Comment:
Is the rounding important or just for ease of reading? My preference is to set `self.bid_rounding_threshold = 2` in the `__init__` method and then use it here. That way it is easier to change. Or we can skip rounding and do it later.

Reply:
Or if rounding is important, add a comment that explains why to keep this at 2.